service/history/workflow/mutable_state_rebuilder.go

```diff
@@ -90,6 +90,36 @@ func NewMutableStateRebuilder(
 }
 
 func (b *MutableStateRebuilderImpl) ApplyEvents(
+	ctx context.Context,
+	namespaceID namespace.ID,
+	requestID string,
+	execution commonpb.WorkflowExecution,
+	history [][]*historypb.HistoryEvent,
+	newRunHistory []*historypb.HistoryEvent,
+) (MutableState, error) {
+	for i := 0; i < len(history)-1; i++ {
+		_, err := b.applyEvents(ctx, namespaceID, requestID, execution, history[i], nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+	newMutableState, err := b.applyEvents(ctx, namespaceID, requestID, execution, history[len(history)-1], newRunHistory)
```

Comment:
Sanity check len(history) - 1?

Reply:
this is correct

the above for loop loops over all event batch except the last one
then use the last event batch and new run event, to generate the new ms iff continue as new

Reply:
I think what YX asking was if `len(history) > 0` is always true. I can't infer that based on the code.

Reply:
i see

input validation should be done at the beginning of the "API", e.g. at frontend service, at network handler.
this piece of logic is internal, so should trust basic input validation is already done.